### PR TITLE
Refactor TokenService#findTokenFromRefreshToken to accept only user-facing refresh tokens

### PR DESCRIPTION
**Pull request by albertzaharovits**
In preparation for the introduction of a new refresh token format,
I've raised this refactoring PR that simplifies the interface of the
`TokenService#findTokenFromRefreshToken` method to only
accept user-facing refresh token arguments.
This reduces token length-based inferences which should simplify
the introduction of a new token format. 

Related ES-6330

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java [hunk 1]
```diff
@@ -102,44 +104,30 @@ private void findAndInvalidateTokens(SamlRealm realm, SamlLogoutRequestHandler.R
...
+                tokenService.invalidateAllTokens(tokens, ActionListener.wrap(tokensInvalidationResult -> {
+                    if (LOGGER.isInfoEnabled() && tokensInvalidationResult.getErrors().isEmpty() == false) {
+                        try (XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent())) {
+                            tokensInvalidationResult.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                            LOGGER.info("Failed to invalidate some SAML access or refresh tokens {}", Strings.toString(builder));
+                        }
+                    }
+                    // return only the total of active tokens for users of the realm, i.e. not the number of actually invalidated tokens
+                    listener.onResponse(tokens.size());
+                }, listener::onFailure));
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java:126]: Here's the core of the change:
Rather than iterating over each of the found tokens (access and refresh), call the `TokenService#invalidateAll` method, which is used in 2 similar other places, and pass all the found tokens for bulk update invalidation.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java [hunk 2]
```diff
@@ -992,35 +977,26 @@ private void findTokenFromRefreshToken(String refreshToken, Iterator<TimeValue>
             );
             findTokenFromRefreshToken(refreshToken, securityMainIndex, backoff, listener);
         } else {
-            if (refreshToken.length() == HASHED_TOKEN_LENGTH) {
-                logger.debug("Assuming a hashed refresh token [{}] retrieved from the tokens index", refreshToken);
-                findTokenFromRefreshToken(refreshToken, securityTokensIndex, backoff, listener);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:997]: And this is the motivation for the change:
The interface of the `findTokenFromRefreshToken` method was complicated because it accepted both "user-facing" refresh tokens (i.e. unhashed) as well as "internal-only" ones (i.e. hashed), and it differentiated between the two by length...
But with the change at https://github.com/elastic/elasticsearch/pull/97422/files#r1254553360 , "internal-only" ones are not a valid input anymore, so this length check can be discarded.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java [hunk 3]
```diff
@@ -992,35 +977,26 @@ private void findTokenFromRefreshToken(String refreshToken, Iterator<TimeValue>
...
-                    logger.debug("Decoded refresh token [{}] with version [{}] is invalid.", unencodedRefreshToken, refreshTokenVersion);
-                    listener.onResponse(SearchHits.EMPTY_WITH_TOTAL_HITS);
-                } else {
-                    // TODO Remove this conditional after backporting to 7.x
-                    if (refreshTokenVersion.onOrAfter(VERSION_HASHED_TOKENS)) {
-                        final String hashedRefreshToken = hashTokenString(unencodedRefreshToken);
-                        findTokenFromRefreshToken(hashedRefreshToken, securityTokensIndex, backoff, listener);
-                    } else {
-                        findTokenFromRefreshToken(unencodedRefreshToken, securityTokensIndex, backoff, listener);
-                    }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:1022]: As an "extra", I've also removed this if-condition branch as it was not viable (commit https://github.com/elastic/elasticsearch/pull/97422/commits/5e7f77a9af29cf87ebbef5d6378d6d0d8de85815): refresh tokens prepended with a version are always hashed (never raw).

Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:1022]: Either way that if-block (as it existed before) doesn't actually make sense to me:

`VERSION_TOKENS_INDEX_INTRODUCED` = `VERSION_HASHED_TOKENS` = `7.2`, so `refreshTokenVersion.onOrAfter(VERSION_HASHED_TOKENS)` was always guaranteed to be true since otherwise, we would have not gotten to the else clause of, i.e.,:

```
if (refreshTokenVersion.before(7.2) || ...) {

} else { 
  // refreshTokenVersion.onOrAfter(7.2) must hold regardless of whether we backport anything or not
} 
```

tl;dr: thanks for removing this 👍

**Review by n1v0lg — COMMENTED** [review]
Looking good, this is a neat simplification! I haven't spotted any issues but want to give this one more pass on Monday since the `TokenService` is... complicated.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java [hunk 4]
```diff
@@ -102,44 +104,30 @@ private void findAndInvalidateTokens(SamlRealm realm, SamlLogoutRequestHandler.R
...
             }
         }, listener::onFailure));
     }
 
-    private void invalidateTokenPair(Tuple<UserToken, String> tokenPair, ActionListener<TokensInvalidationResult> listener) {
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java:124]: Only an observation: before this change, we would invalidate tokens in pairs (refresh_0, access_0), ..., (refresh_n, access_n). Now we invalidate all refresh tokens first, then all access tokens. I can't think of why that'd be a problem and in fact it's probably desirable. Just jotting this down as a diff in behavior.

**Review by n1v0lg — APPROVED** [review]
LGTM! Neat simplification for sure, no blocking comments.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java (hunk 3 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:1022]: Yes, the else branch was not possible, as you describe, there are multiple version constants equaling `7.2`, introduced in independent work streams for token format changes that got merged in the same release, and so it's not possible that one be false and the other true.

> This is pertaining to the "// TODO Remove this conditional after backporting to 7.x" and the functionality that was introduced by https://github.com/elastic/elasticsearch/pull/54532 and backported to 7.x with https://github.com/elastic/elasticsearch/commit/ac87c100396992a1d96954739be7e8569b40bff5, correct?

I assumed as much myself, I don't recall anything about the context of the comment...

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java [hunk 5]
```diff
@@ -734,7 +719,7 @@ public void invalidateActiveTokensForRealmAndUser(
      * @param tokenTuples The user token tuples for which access and refresh tokens (if exist) should be invalidated
      * @param listener  the listener to notify upon completion
      */
-    private void invalidateAllTokens(Collection<Tuple<UserToken, String>> tokenTuples, ActionListener<TokensInvalidationResult> listener) {
+    public void invalidateAllTokens(Collection<Tuple<UserToken, String>> tokenTuples, ActionListener<TokensInvalidationResult> listener) {
         maybeStartTokenRemover();
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:723]: Good point!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java [hunk 6]
```diff
@@ -102,44 +104,30 @@ private void findAndInvalidateTokens(SamlRealm realm, SamlLogoutRequestHandler.R
...
-                GroupedActionListener<TokensInvalidationResult> groupedListener = new GroupedActionListener<>(
-                    tokens.size(),
-                    ActionListener.wrap(collection -> listener.onResponse(collection.size()), listener::onFailure)
-                );
-                tokens.forEach(tuple -> invalidateTokenPair(tuple, groupedListener));
+                tokenService.invalidateAllTokens(tokens, ActionListener.wrap(tokensInvalidationResult -> {
+                    if (LOGGER.isInfoEnabled() && tokensInvalidationResult.getErrors().isEmpty() == false) {
+                        try (XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent())) {
+                            tokensInvalidationResult.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                            LOGGER.info("Failed to invalidate some SAML access or refresh tokens {}", Strings.toString(builder));
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java:121]: I opted to keep the same log level that's been used in `invalidateAccessToken`.
If this was a case of introducing new logging, I would've probably opted for WARN myself too, but it makes little difference since both WARN and INFO levels are enabled by default.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java (hunk 4 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java:124]: Yes, I'm aware of this. I couldn't think of any reason why this change would matter in practice.

**Comment by albertzaharovits** [comment]
Thank you for the quick review @n1v0lg !
